### ISSUE TITLE
If/else condition based on browser from projects in config (JS/TS)

### ISSUE THREAD

**Issue by MAlterCVET** [microsoft/playwright#19564]
I am currently writing UI tests and I have three projects in my playwright.config.ts file as per [this doc](https://playwright.dev/docs/api/class-testproject) for cross-browser testing. One of them (WebKit) is slower than the others and is causing tests to fail from exceeded wait times where the other browsers do not.

I've taken some of the structure and functions of my code from [this demo](https://github.com/JayKishoreDuvvuri/Playwright-JavaScript-SauceDemo), including the `.wait()` function in the basePage class, but I've modified it to be environment-specific:
```
async wait() {
    await this.page.waitForTimeout(DATA.waitTime);
}
```
(where `DATA` refers to a config file found by .env)
What I'd like to do is add a condition to this function to increase the wait time if Playwright is using the WebKit project, as in:
```
async wait() {
   if (/*browser is webkit*/) {
       await this.page.waitForTimeout(DATA.waitTime * 2);
   } else {
       await this.page.waitForTimeout(DATA.waitTime);
   }
}
```
Is this possible? If so, what code should I use in that if/else condition? If not, how else can I increase the `waitForTimeout();` time for one project but not the others?

**Comment by mxschmitt** [comment]
You should never use waitForTimeout inside a test. Its not predicable how long you should wait, sometimes when the CPU is overloaded, different browsers can be faster than others.

In which cases do you have waitForTimeouts inside your codebase? Playwright does auto-waiting by default for click, fill etc., only when you are not using the [web-first assertions](https://playwright.dev/docs/test-assertions), such flaky alternatives need to be considered.

**Comment by MAlterCVET** [comment]
Believe me, I wish I could avoid it. Like in the demo project, I usually use other methods such as `waitForSelector()` when I can, but I've had to use `waitForTimeout()` in cases when there's no way to tell what the test should be waiting for. In my case, working on an e-commerce site, some elements are so context-dependent and changeable that it's counterproductive to expect a specific one. Plus our subprod environments are laggy at best, which is why I've assigned different wait times to different environments.

**Comment by pavelfeldman** [comment]
Sadly, you can't expect the tests to pass if you are waiting for time. Time is not a concept on server as well, if you ever want to run it on CI. If you are lacking the UI signals that are handled by locators (btw, you _never_ need `waitForSelector()` in Playwright), you can try waiting for the network signals:

```js
const responsePromise = page.waitForResponse('**/api/fetch_data');
await page.getByText('Update').click();
const response = await responsePromise;
```

There isn't much happening in the web page usually and the combination of UI + network hooks pretty much always solves it.

**Comment by MAlterCVET** [comment]
Ignoring the original purpose, is there a way to write an if/else condition to check the browser independent of the `test` object? There are definitely other situations when that would be useful.

**Comment by pavelfeldman** [comment]
```js
test('test', async ({ browserName }) => {
  if (browserName === 'webkit') ...
});

**Comment by MAlterCVET** [comment]
That looks perfect... almost. Is `browserName` an import from @playwright/test? Also, will Playwright let me put that "test" into an imported function that's used in another test? If not, I'll have to change a lot of my tests to accommodate an extra parameter.

**Comment by pavelfeldman** [comment]
> Is browserName an import from @playwright/test

It is not, check out this doc: https://playwright.dev/docs/next/api/class-fixtures#fixtures-browser-name. It is basically passed to every test, if you want to pass browserName into a helper function, you would need to manually plumb it there.

**Comment by pavelfeldman** [comment]
Closing as per above, please feel free to open a new issue if this does not cover your use case.
